doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc: Execute the Program with Multi Thread
+
+## Abstract
+
+This Design Doc propose an approach to make the user-defined Op graph
+running with multi threads, we will use an auto transpiler to convert the user-defined
+Op graph to a multi-thread Op graph, and run `ParallelDo` Op to run the
+multi-thread graph.
+
+## Graph Converter
+<img src="src/multi-threads/single-thread@3x.png" width="300">
+
+After converted:
+
+<img src="src/multi-threads/multi-threads@3x.png" width="1000">
+
+## Implement
+
+- `Multi-Thread Transpiler`  will convert the graph to a multi-threads graph
+    which would be executed with multi-threads.
+- `BlockingCounter` will `Init/Decrement` a condition variable, and Blocking `Wait`
+    for the condition variable become `0`:
+    ```cpp
+    BlockingCounter bc(thread_count);
```

Comment:
Is it same with `WaitGroup`?

Reply:
Yes, they are the same, WaitGroup is [here](https://golang.org/pkg/sync/#WaitGroup)